### ISSUE TITLE
Different variance detection between Python versions 3.12 and 3.13

### ISSUE THREAD

**Issue by alexandermalyga** [microsoft/pyright#10464]
Pyright is detecting the type parameter `ModelT@Event` as invariant while using Python 3.13, and as covariant in 3.12. 

```python
from dataclasses import dataclass
from typing import Self
from abc import ABC, abstractmethod

@dataclass(frozen=True)
class Event[ModelT: "Model"]:
    model_type: type[ModelT]

class Model(ABC):

    @abstractmethod
    def event(self) -> Event[Self]: ...

@dataclass(frozen=True)
class MyEvent(Event["MyModel"]):
    ...

class MyModel(Model):

    def event(self) -> MyEvent:
        return MyEvent(model_type=type(self))
```

This produces the following error with Python 3.13, while not producing errors in 3.12:

```
error: Method "event" overrides class "Model" in an incompatible manner
    Return type mismatch: base method returns type "Event[Model]", override returns type "MyEvent"
      "MyEvent" is not assignable to "Event[Model]"
        Type parameter "ModelT@Event" is invariant, but "MyModel" is not the same as "Model" (reportIncompatibleMethodOverride)
```

I think it's related to `frozen=True`, because with that parameter set to `False` in Python 3.12 Pyright detects the type var as invariant.
I'm using the Pyright CLI version `1.1.400`.

**Comment by erictraut** [comment]
Pyright is working correctly here, so I don't consider this a bug.

This is because of a new dataclass method named `__replace__` that the runtime automatically generates for non-frozen dataclasses starting in Python 3.13. The signature of `__replace__` forces invariance because it uses the field type in a contravariant position.

There have been some suggestions that the typing spec should perhaps be amended to exempt `__replace__` from variance calculations and override checks, but these haven't led to any formal proposals. It's not clear whether the typing community would get behind such a change because it's technically incorrect from a type theory standpoint and could lead to unsoundness. However, one may be able to make a compelling argument on the grounds of practicality and utility.

If you'd like to propose something here, I recommend starting a discussion in the [Python typing forum](https://discuss.python.org/c/typing/32). [Here](https://github.com/python/typing-council#decisions) is the overview of how to amend the typing spec.

In the meantime, I'm going to close this issue because pyright is working correctly according to the current typing spec.
